### PR TITLE
[HEVCe] Enable SCC on hevc encoder

### PR DESCRIPTION
**Pull request by JeremyShang**

### CODE REVIEW

**Comment by onabiull** [comment]
Could you please also update doc/mediasdk-man.md to reflect how to use that functionality?

File: _studio/mfx_lib/encode_hw/hevc/linux/g12/hevcehw_g12_scc_lin.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    void SubmitTask(const FeatureBlocks& /*blocks*/, TPushST Push) override
+    {
+        Push(BLK_PatchDDITask
+            , [this](StorageW& global, StorageW& /*s_task*/) -> mfxStatus
+        {
+            MFX_CHECK(m_bPatchNextDDITask || m_bPatchDDISlices, MFX_ERR_NONE);
+            auto& ddiPar = HEVCEHW::Gen12::Glob::DDI_SubmitParam::Get(global);
+            auto  itPPS  = std::find_if(std::begin(ddiPar), std::end(ddiPar)
+                , [](HEVCEHW::Base::DDIExecParam& ep) { return (ep.Function == VAEncPictureParameterBufferType); });
+            MFX_CHECK(itPPS != std::end(ddiPar) && itPPS->In.pData, MFX_ERR_UNKNOWN);
```
Comment (lakulako) [inline _studio/mfx_lib/encode_hw/hevc/linux/g12/hevcehw_g12_scc_lin.h:62]: Please don't use MFX_ERR_UNKNOWN where is possible to guess what is going on.

File: _studio/mfx_lib/encode_hw/hevc/agnostic/g12/hevcehw_g12_scc.cpp [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        else
+        {
+            ssh.num_ref_idx_l0_active_minus1++; //for cur pic ref
+        }
+
+        return MFX_ERR_NONE;
+    });
+}
+
+#endif //defined(MFX_ENABLE_H265_VIDEO_ENCODE)
```
Comment (lakulako) [inline _studio/mfx_lib/encode_hw/hevc/agnostic/g12/hevcehw_g12_scc.cpp:346]: Please add \n at EOF

**Comment by vilichev** [comment]
> Could you please also update doc/mediasdk-man.md to reflect how to use that functionality?

@onabiull current doc/mediasdk-man.md already have mentions about MFX_PROFILE_HEVC_SCC (see CodecProfile section). So I think it already covers this functionality

File: _studio/mfx_lib/encode_hw/hevc/linux/g12/hevcehw_g12_scc_lin.h (hunk 1 above)
Comment (JeremyShang) [reply _studio/mfx_lib/encode_hw/hevc/linux/g12/hevcehw_g12_scc_lin.h]: Thanks, changed this error message to  MFX_ERR_NOT_FOUND.

File: _studio/mfx_lib/encode_hw/hevc/agnostic/g12/hevcehw_g12_scc.cpp (hunk 2 above)
Comment (JeremyShang) [reply _studio/mfx_lib/encode_hw/hevc/agnostic/g12/hevcehw_g12_scc.cpp]: Thanks, done.

File: _studio/mfx_lib/encode_hw/hevc/linux/g12/hevcehw_g12_scc_lin.h [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#pragma once
+
+#include "mfx_common.h"
+#if defined(MFX_ENABLE_HEVCE_SCC) && defined (MFX_VA_LINUX)
```
Comment (dvrogozh) [inline _studio/mfx_lib/encode_hw/hevc/linux/g12/hevcehw_g12_scc_lin.h:24]: I don't see you define `MFX_ENABLE_HEVCE_SCC` anywhere in this code. And it does not present in the current code base either:
```
$ fgrep -rsn MFX_ENABLE_HEVCE_SCC _studio/
$
```

This makes me think you submit non-verified change. Do you have sample_encode command line to show this change is functional?

Comment (JeremyShang) [reply _studio/mfx_lib/encode_hw/hevc/linux/g12/hevcehw_g12_scc_lin.h]: Hi. Thanks for pointing out. Sorry, I made a mistake regarding code base which introduces this macro. Have removed it. Yes, I have a command line proved functional with the latest commit. Command Line:  
 ./sample_encode h265 -i in.yuv -o out.bin  -qpi 17 -qpp 17 -qpb 17 -qsv-ff -hw -f 30 -w 480 -h 272 -CodecProfile 9 -g 10000 -r 1 -u 4 -n 20

**Comment by JeremyShang** [comment]
@vilichev  Hi. Do you suggest we need a macro for HEVCe SCC, like 'MFX_ENABLE_HEVCE_SCC'?

**Comment by vilichev** [comment]
> Hi. Do you suggest we need a macro for HEVCe SCC, like 'MFX_ENABLE_HEVCE_SCC'?
I don't know cases when we will need such macro to disable scc

**Comment by dvrogozh** [comment]
Year, less number of macro defs we will have - the better. No need to be too detailed especially if there is no much sense to disable the feature. It won't save us anything I guess.

**Comment by dvrogozh** [comment]
Might require rebase to avoid merge commit.

**Comment by JeremyShang** [comment]
> Might require rebase to avoid merge commit.

Thanks, rebase done.

File: _studio/mfx_lib/encode_hw/hevc/agnostic/g12/hevcehw_g12_scc.cpp [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+    return true;
+}
+
+void SCC::Query1NoCaps(const FeatureBlocks& /*blocks*/, TPushQ1 Push)
+{
+    Push(BLK_SetCallChains
+        , [this](const mfxVideoParam& par, mfxVideoParam&, StorageRW& strg) -> mfxStatus
+    {
+        MFX_CHECK(par.mfx.CodecProfile == MFX_PROFILE_HEVC_SCC, MFX_ERR_NONE);
+        MFX_CHECK(!IsOff(par.mfx.LowPower), MFX_ERR_NONE);
```
Comment (JeremyShang) [inline _studio/mfx_lib/encode_hw/hevc/agnostic/g12/hevcehw_g12_scc.cpp:142]: Update Checking Logic here. If -qsv ff isn't set by user , lib throws unsupported with scc profile in Query.

**Pull request merged**